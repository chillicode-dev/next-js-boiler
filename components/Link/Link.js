// Vendor
import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import NextLink from 'next/link';
import {useRouter} from 'next/router';
// Internals
import {url, env} from 'utils';

const Link = ({className, activeClassName, children, href, external, targetBlank, protocol, htmlProps, ...props}) => {
  const router = env.isTest() ? {pathname: '/'} : useRouter();
  const classes = cn({
    [activeClassName]: activeClassName && router.pathname === href,
    [className]: className,
  });
  const nativeHtmlProps = {
    className: classes,
    ...htmlProps,
  };
  const isExternal = external || url.isAbsolute(href);

  // Return plain <a> if detected that link is external or special protocol is specified
  if (isExternal || protocol) {
    nativeHtmlProps.href = href;
    if (targetBlank) {
      nativeHtmlProps.target = '_blank';
      nativeHtmlProps.rel = 'nofollow noopener';
    }
    if (protocol) {
      nativeHtmlProps.href = url.changeProtocol(nativeHtmlProps.href, protocol);
    }
    return <a {...nativeHtmlProps}>{children}</a>;
  }

  // Return next/link if link is internal
  return (
    <NextLink href={href} {...props}>
      <a {...nativeHtmlProps}>{children}</a>
    </NextLink>
  );
};

Link.defaultProps = {
  className: '',
  activeClassName: '',
  external: false,
  targetBlank: false,
  protocol: '',
  htmlProps: {},
};

Link.propTypes = {
  children: PropTypes.node.isRequired,
  href: PropTypes.oneOfType([PropTypes.string, PropTypes.object]).isRequired,
  className: PropTypes.string,
  activeClassName: PropTypes.string,
  external: PropTypes.bool,
  targetBlank: PropTypes.bool,
  protocol: PropTypes.string,
  htmlProps: PropTypes.object,
};

export default Link;
