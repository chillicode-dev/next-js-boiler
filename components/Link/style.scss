.Link {
  @include link($linkColor, $linkColorActive);
}
