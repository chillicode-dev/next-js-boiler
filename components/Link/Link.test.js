import React from 'react';
import renderer from 'react-test-renderer';
import Link from './Link';

it('Renders Link correctly', () => {
  const elements = (
    <ul>
      <li>
        <Link href="/" activeClassName="active">
          Main page
        </Link>
      </li>
      <li>
        <Link href="/about" activeClassName="active">
          About
        </Link>
      </li>
      <li>
        <Link href="//ya.ru">External</Link>
      </li>
      <li>
        <Link href="//ya.ru" targetBlank>
          External with targetBlank
        </Link>
      </li>
      <li>
        <Link href="+79993335544" protocol="tel">
          Call me
        </Link>
      </li>
      <li>
        <Link href="adlite@yandex.ru" protocol="mailto">
          Send me email
        </Link>
      </li>
    </ul>
  );
  const tree = renderer.create(elements).toJSON();
  expect(tree).toMatchSnapshot();
});
