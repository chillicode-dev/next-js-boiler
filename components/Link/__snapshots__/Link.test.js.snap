// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Link correctly 1`] = `
<ul>
  <li>
    <a
      className="active"
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Main page
    </a>
  </li>
  <li>
    <a
      className=""
      href="/about"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      About
    </a>
  </li>
  <li>
    <a
      className=""
      href="//ya.ru"
    >
      External
    </a>
  </li>
  <li>
    <a
      className=""
      href="//ya.ru"
      rel="nofollow noopener"
      target="_blank"
    >
      External with targetBlank
    </a>
  </li>
  <li>
    <a
      className=""
      href="tel://+79993335544"
    >
      Call me
    </a>
  </li>
  <li>
    <a
      className=""
      href="mailto://adlite@yandex.ru"
    >
      Send me email
    </a>
  </li>
</ul>
`;
